torch/distributed/_sharding_spec/api.py

```diff
@@ -96,3 +96,173 @@ def placement(self) -> ShardPlacement:
         Retrieves the shard placement.
         """
         return self._placement
+
+class GenericShardingSpec(PlacementSpec):
+
+    class Shard:
```

Comment:
Since there can be a large number of `Shard` instance, better to add `__slots__` for memory optimization.